src/vulkan-renderer/io/byte_stream.cpp

```diff
@@ -0,0 +1,112 @@
+#include "inexor/vulkan-renderer/io/byte_stream.hpp"
+#include "inexor/vulkan-renderer/world/cube.hpp"
+
+#include <fstream>
+
+namespace inexor::vulkan_renderer::io {
+std::vector<std::uint8_t> ByteStream::read_file(const std::filesystem::path &path) {
+    std::ifstream stream(path, std::ios::in | std::ios::binary);
+    return {std::istreambuf_iterator<char>(stream), std::istreambuf_iterator<char>()};
+}
+
+ByteStream::ByteStream(std::vector<std::uint8_t> buffer) : m_buffer(std::move(buffer)) {}
+
+ByteStream::ByteStream(const std::filesystem::path &path) : ByteStream(read_file(path)) {}
+
+std::size_t ByteStream::size() const {
+    return m_buffer.size();
+}
+[[nodiscard]] const std::vector<std::uint8_t> &ByteStream::buffer() const {
+    return m_buffer;
+}
+
+void ByteStreamReader::check_end(const std::size_t size) const {
+    if (std::distance(m_iter, m_stream.buffer().end()) < size) {
+        throw std::runtime_error("end would be overrun");
+    }
+}
+
+ByteStreamReader::ByteStreamReader(const ByteStream &stream) : m_stream(stream), m_iter(stream.buffer().begin()) {}
+
+void ByteStreamReader::skip(const std::size_t size) {
+    std::size_t skip = std::min(
+        size, std::size_t(std::distance<std::vector<std::uint8_t>::const_iterator>(m_iter, m_stream.buffer().end())));
+    std::advance(m_iter, skip);
+}
+
+std::size_t ByteStreamReader::remaining() const {
+    return std::distance<std::vector<std::uint8_t>::const_iterator>(m_iter, m_stream.buffer().end());
+}
+
+template <>
+std::uint8_t ByteStreamReader::read() {
+    check_end(1);
+    return *m_iter++;
+}
+
+template <>
+std::uint32_t ByteStreamReader::read() {
+    check_end(4);
+    return *m_iter++ | (*(m_iter++) << 8) | (*(m_iter++) << 16) | (*(m_iter++) << 24);
+}
+
+template <>
+std::string ByteStreamReader::read(const std::size_t &size) {
+    check_end(size);
+    auto start = m_iter;
+    std::advance(m_iter, size);
+    return std::string(start, m_iter);
+}
+
+template <>
+world::Cube::Type ByteStreamReader::read() {
+    return world::Cube::Type(read<std::uint8_t>());
+}
+
+template <>
+std::array<world::Indentation, 12> ByteStreamReader::read() {
+    check_end(9);
+    std::array<world::Indentation, 12> indentations;
+    auto writer = indentations.begin();
+    const auto end = m_iter + 9;
+    while (m_iter != end) {
+        *writer++ = std::move(world::Indentation(*m_iter >> 2));
+        *writer++ = std::move(world::Indentation(((*m_iter & 0b00000011) << 4) | (*(++m_iter) >> 4)));
+        *writer++ = std::move(world::Indentation(((*m_iter & 0b00001111) << 2) | (*(++m_iter) >> 6)));
+        *writer++ = std::move(world::Indentation(*(m_iter++) & 0b00111111));
```

Comment:
```suggestion
        *writer++ = world::Indentation(*m_iter >> 2);
        *writer++ = world::Indentation(((*m_iter & 0b00000011) << 4) | (*(++m_iter) >> 4));
        *writer++ = world::Indentation(((*m_iter & 0b00001111) << 2) | (*(++m_iter) >> 6));
        *writer++ = world::Indentation(*(m_iter++) & 0b00111111);
```
`std::move()` has no effect here, since they're all temporaries.